File: packages/flutter/lib/src/device_context.dart
import 'dart:io';
import 'dart:ui' as ui;

import 'package:device_info_plus/device_info_plus.dart';
import 'package:package_info_plus/package_info_plus.dart';

import 'config.dart';

/// Snapshots device + app context for inclusion in every report. Mirrors the
/// iOS/Android `DeviceContext` so the cross-platform schema stays identical.
class DeviceContext {
  static Future<Map<String, dynamic>> capture() async {
    final pkg = await PackageInfo.fromPlatform();
    final info = DeviceInfoPlugin();
    final view = ui.PlatformDispatcher.instance.implicitView;

    final deviceMap = <String, dynamic>{};
    if (Platform.isAndroid) {
      final a = await info.androidInfo;
      deviceMap.addAll({
        'manufacturer': a.manufacturer,
        'model': a.model,
        'osName': 'Android',
        'osVersion': a.version.release,
        'apiLevel': a.version.sdkInt,
      });
    } else if (Platform.isIOS) {
      final i = await info.iosInfo;
      deviceMap.addAll({
        'manufacturer': 'Apple',
        'model': i.utsname.machine,
        'osName': i.systemName,
        'osVersion': i.systemVersion,
      });
    } else {
      deviceMap.addAll({
        'osName': Platform.operatingSystem,
        'osVersion': Platform.operatingSystemVersion,
      });
    }
    if (view != null) {
      final size = view.physicalSize;
      deviceMap['screenWidth'] = size.width;
      deviceMap['screenHeight'] = size.height;
      deviceMap['devicePixelRatio'] = view.devicePixelRatio;
    }

    return <String, dynamic>{
      'platform': Platform.operatingSystem,
      'sdkName': MushiInfo.sdkName,
      'sdkVersion': MushiInfo.sdkVersion,
      'timestamp': DateTime.now().toUtc().millisecondsSinceEpoch,
      'locale': ui.PlatformDispatcher.instance.locale.toLanguageTag(),
      'app': {
        'id': pkg.packageName,
        'version': pkg.version,
        'build': pkg.buildNumber,
      },
      'device': deviceMap,
    };
  }
}

File: packages/android/src/main/kotlin/dev/mushimushi/sdk/capture/DeviceContext.kt
package dev.mushimushi.sdk.capture

import android.content.Context
import android.os.Build
import dev.mushimushi.sdk.MushiInfo
import java.util.Locale
import java.util.TimeZone

/**
 * Snapshots device + app context for inclusion in every report. Mirrors
 * the iOS DeviceContext so the cross-platform schema stays identical.
 */
object DeviceContext {
    fun capture(context: Context): Map<String, Any?> {
        val app = context.applicationContext
        val pm = app.packageManager
        val pkg = app.packageName
        val info = runCatching { pm.getPackageInfo(pkg, 0) }.getOrNull()
        val metrics = app.resources.displayMetrics

        return mapOf(
            "platform" to "android",
            "sdkName" to MushiInfo.SDK_NAME,
            "sdkVersion" to MushiInfo.SDK_VERSION,
            "timestamp" to System.currentTimeMillis(),
            "locale" to Locale.getDefault().toLanguageTag(),
            "timezone" to TimeZone.getDefault().id,
            "app" to mapOf(
                "id" to pkg,
                "version" to info?.versionName,
                "build" to (if (Build.VERSION.SDK_INT >= 28) info?.longVersionCode else info?.versionCode?.toLong())
            ),
            "device" to mapOf(
                "manufacturer" to Build.MANUFACTURER,
                "model" to Build.MODEL,
                "osName" to "Android",
                "osVersion" to Build.VERSION.RELEASE,
                "apiLevel" to Build.VERSION.SDK_INT,
                "screenWidth" to metrics.widthPixels,
                "screenHeight" to metrics.heightPixels,
                "density" to metrics.density
            )
        )
    }
}
